README.MD
Tezos-rs
===========

The purpose of this project is to implement a secure, trustworthy, open-source Tezos node in Rust. In addition to implementing a new node, the project seeks to maintain and improve the Tezos node wherever possible. 

The project can be considered as a proof of concept. This PoC demonstrates the viability of a node build that uses Rust. 

## Components

The current PoC of the Tezos node consists of multiple components which are yet to be fully implemented:

* **p2p/client:** This is the p2p component responsible for receiving and sending messages within Tezos network. Internally, p2p allocates a component, called a p2p-peer, for each node within the reachable network. The p2p-peer listens to incoming messages from peers and routes requests to individual peers. The p2p/client uses separate components for serialization/deserialization of messages between Tezos message format and Rust structures.

* **data_encoding/de:** All incoming messages are transformed into standard Rust structures for easy manipulation using de component. This component implements serialization and deserialization of all data types used in Tezos messages.

* **crypto/box:** Component contains cryptographic algorithms for encryption and decryption of messages

* **storage/db:** In the current phase (PoC), this is a very simplistic in memory DB for storing information about the current state of the blockchain.

* **rpc/server:** the component used for the implementation of tezos-rs REST API. Using the nodes REST api, the user is able to get insights about the current state of the node as well as execute commands on the node

The relations between individual components are described in the diagram depicted below.

![Preview1](./docs/images/class_diagram.png)


# How to run the node

Requirements
------------

**1. Install Rust** 

We recommend installing Rust through rustup.

Run the following in your terminal, then follow the onscreen instructions.

```
curl https://sh.rustup.rs -sSf | sh
```

**2. Rust version** 

Rust nightly is required to build this project.
```
rustup install nightly-2019-03-14
rustup default nightly-2019-03-14
```
Application has been tested to compile with `rustc 1.35.0-nightly (bc44841ad 2019-03-14)`.


Building
--------

**3. Build node** 

On linux systems:

```
export SODIUM_USE_PKG_CONFIG=1
cargo build
```

On windows systems with vcpkg:
```
set SODIUM_USE_PKG_CONFIG=1
set VCPKGRS_DYNAMIC=1
cargo build
```

Run
------------
**4. Run node** 

```
cargo run  
```

![Preview1](./docs/images/bash_cargo_run.gif)


# RPC API

#### /network/points
Returns connected peers, after successful authentication.

```
curl http://127.0.0.1:18732/network/points
```

![Preview1](./docs/images/bash_network_points.gif)


#### /chains/main/blocks/head
Returns current head.

```
curl http://127.0.0.1:18732/chains/main/blocks/head
```

![Preview1](./docs/images/bash_chains_main_blocks_head.gif)

# Run configuration

**RPC port**
```
--rpc-port 8732
```

**P2P port**
```
--p2p-port 9732
```

**Bootstrap peers**
```
--peers 127.0.0.1:7777,127.0.0.1:8888
```

**Bootstrap address**

```
--bootstrap-lookup-address boot.tzalpha.net,bootalpha.tzbeta.net
```

**Location to identity.json file**
```
--identity /opt/tezos-env/rust-node/identity.json
```

**Disable JSON messages logging**
```
--logj false
```

**Disable HEX messages logging**
```
--logh false
```


e.g.:
```
cargo run -- --rpc-port 9998 --p2p-port 7533 --identity ./config/identity.json --logj true --logh false
```


# What node can currently do

The two diagrams shown below depict the interaction between objects in a sequential order. The sequence of actions is described in a short sentence written above each diagram.

#### 1. Connect to the Tezos p2p network and listen for incoming connections

![Preview2](./docs/images/bootstrap.png)


#### 2. Return the current head of chain

![Preview4](./docs/images/get_current_head.png)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [Unreleased]

### Added

- Nothing.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

### Security

- Nothing.

### Performance

- Nothing.

## [3.1.1] - 2022-06-28

### Fixed

- Added missing jakarta protocol recognition to context timings tool.

## [3.1.0] - 2022-06-27

### Added
- Jakarta support in rewards RPC.

### Fixed

- Fixed rewards RPC edge cases when the interrogated cycle was lower than preserved_cycles.
- Fixed embedded baker for blocks with predecessor containing a protocol activation.
- Minor fixes in Tezos encoding library.

### Security

- Removed `chrono` dependency from `node_monitoring` application.

## [3.0.0] - 2022-06-25

### Added
- Internal baker built inside the node.
- Prechecking for manager operations.
- Jakarta support in mempool and bakers.

### Changed

- New version of block application code in the protocol runner that performs extra checks.
- Prechecking for endorsement operations is now disabled by default.

## [2.3.6] - 2022-06-16

### Changed

- Don't disconnect peers when in private mode.
- Include `tezedge-baker` program in docker image.

## [2.3.5] - 2022-06-16

### Fixed

- Baker live_block filter.

## [2.3.4] - 2022-06-15

### Added

- Double-baking and double-endorsement protection in baker
- Updated documentation.

## [2.3.3] - 2022-06-10

### Fixed

- Fixed possible errors while calling RPCs due to the WebSocket RPC feature.
- Fixed handling of lazy encoded micheline values in operations.
- Fixed unimplemented panic in describe endpoint.

### Performance

- Split the rewards RPC into two (/dev/rewards/cycle/:cycle_num and /dev/rewards/cycle/:cycle_num/:delegate) to further optimise performance

## [2.3.2] - 2022-06-07

### Fixed

- In the baker, filter out outdated operations when injectiong a block.
- Better handling of injection RPC responses for unparseable operations.

## [2.3.1] - 2022-06-06

### Added

- New baker documentation.

### Fixed

- Handle corner case when recovering the context storage from a hard crash.
- Add missing flush call for the chain storage after block application to ensure it can be recovered from a hard crash.
- Bug in baker scheduler that caused some requests to wrongly eclipse previous ones.

## [2.3.0] - 2022-06-03

### Added

- Preliminary support for protocol 013-Jakarta.
- Added a baker implementation.
- Added RPC for baker rewards distribution.
- Added support for tezedge RPCs through websocket using JSON-RPC 2.0.

### Changed

- Changed the monitoring websocket communication from a push based architecture to a request/response architecture.
- Reworked the prevalidated code in the protocol runner.

### Fixed

- `/chains/:chain/blocks` RPC.

## [2.2.0] - 2022-04-29

### Added

- In-memory context backend now regularly takes a snapshot of the most recent commits,
  and is able to restore its state when the node is restarted.
- Prevalidation priority logic, so that injected and consensus operations
  get validated first to increase the chance of them being included
  in the next block.
- Better and more detailed action statistics: https://tezedge.com/#/resources/state
- Slack alerts for missed block/endorsement for a specific baker.
- Report generation in case of missed block/endorsement, which includes
  debug information.

### Fixed

- Issues with mempool and prevalidator.
- Assertion failure in tezedge native context storage when running ithacanet.
- Rendering of operations metadata JSON in RPCs when there is too much data

### Performance

- Optimized block application to avoid reapplying block in case of branch change.

## [2.1.0] - 2022-04-06

### Added

- Support for preendorsements UI.
- Bounds for operations in mempool.

### Fixed

- Fixed issue with node being disconnected from all peers.
- Fixed support for Ithaca endorsements UI.
- Fixed wrong duration in action statistics.

## [2.0.0] - 2022-03-31

### Changed

- Adapt bootstrapping and mempool logic for Ithaca.
- State machine fuzzing improvements.

### Fixed

- On injection RPCs, respect the `is_async` parameter and respond without waiting for the injection to complete.
- Fixed sporadic connectivity issues between the shell and protocol runner processes.

## [1.19.0] - 2022-03-25

### Added

- Import snapshot subcommand for importing snapshots from a remote file server.

### Changed

- The --config-file argument now defaults to the default config file.

## [1.18.1] - 2022-03-22

### Changed

- For macOS-arm64, use an updated libtezos.

## [1.18.0] - 2022-03-22

### Added

- Support for macOS-arm64 (M1 CPUs).
- Block headers download progress logging.

## [1.17.0] - 2022-03-18

### Added

- Preliminary support for protocol 012 (Ithaca 2).
- Re-implemented bootstrapping logic in the state machine.
- In-memory context context now supports loading it's state from disk.
- Disk usage measurements for the context stats db.
- Implemented Octez `tezos-accuser` support.

### Changed

- In-memory context storage now has a new garbage collector that collects everything that doesn't belong to the last few levels.
- Block application retry logic is now more alike Octez, and can handle protocol runner failures.

### Fixed

- Fixed `/monitoring/valid_blocks` RPC.
- Cache-related block application retry logic now in synch with the one in Octez.

## [1.16.1] - 2022-03-04

### Changed

- Rust 2021 edition is used.
- The option `--disable-apply-retry` now is `false` by default.

### Fixed

- Fixed `/chains/:chain-id:/blocks` RPC that prevented seed nonces to be revealed.
- Fixed potential node stucking when using `--disable-apply-retry=true`.
- Fixed issue that caused the snapshot command to sometimes timeout on slower computers.

### Security

- RUSTSEC-2020-0071 is fixed by using newer `time` crate version.

run.sh
#!/usr/bin/env bash
# Copyright (c) SimpleStaking and Tezedge Contributors
# SPDX-License-Identifier: MIT

# stop on first error
set -e

export RUSTC_TOOLCHAIN_VERSION="1.58.1"

warn_if_not_using_recommended_rust() {
  EXPECTED_RUSTC_VERSION=$RUSTC_TOOLCHAIN_VERSION
  RUSTC_VERSION=$(rustc --version)

  # If longer than 8 chars then it is a date, handle accordingly.
  # The date from the result of `rustc --version` for nightly versions is one day before
  if [[ "${#RUSTC_TOOLCHAIN_VERSION}" -gt 8 ]]; then
    EXPECTED_RUSTC_VERSION=$(date -d "$RUSTC_TOOLCHAIN_VERSION -1 day" +"%Y-%m-%d")
  fi

  if [[ "${RUSTC_VERSION}" != *"${EXPECTED_RUSTC_VERSION}"* ]]; then
    echo -e "\e[33mWarning:\e[0m please use supported rust toolchain version \e[97m$RUSTC_TOOLCHAIN_VERSION\e[0m."
    echo    "Your current version $RUSTC_VERSION."
    echo    "See project README.md on how to install a supported rust toolchain."
  fi
}

help() {
  echo -e "Usage: \e[97m$0\e[0m \e[32mMODE\e[0m"
  echo -e "Possible values for \e[32mMODE\e[0m are:"
  echo -e " \e[32mnode\e[0m       Start tezedge node. You can specify additional arguments here."
  echo    "              # $0 node -v"
  echo -e " \e[32mrelease\e[0m    Start tezedge node in release mode. You can specify additional arguments here."
  echo    "            Release mode means that the node binary is compiled with '--release' flag."
  echo    "              # $0 release -v"
  echo -e " \e[32mdocker\e[0m     Run tezedge node as a docker container. It is possible to specify additional commandline arguments."
  echo    "              # $0 docker -t 4 -T 12"
  echo    "              # $0 docker --help"
  echo -e " \e[32mfuzz\e[0m       Start tezedge node in fuzz mode. You can specify additional arguments here."
  echo    "            Fuzz mode means that the node binary is compiled with most debug options but enabling optimizations."
}

configure_env_variables() {
  options=$(getopt --longoptions debug,release,fuzz,addrsanitizer --options "" --name $0 -- "$@")
  eval set -- "$options"
  while true ; do
    case $1 in
      --debug)
        export PROFILE="debug"
        export CARGO_PROFILE_ARG=""
        shift
        ;;
      --release)
        export PROFILE="release"
        export CARGO_PROFILE_ARG="--release"
        shift
        ;;
      --fuzz)
        export RUSTC_TOOLCHAIN_VERSION="2021-12-22"
        export RUSTFLAGS="-Zprofile -C force-frame-pointers --cfg dyncov"
        export PROFILE="fuzz"
        export CARGO_PROFILE_ARG="-Z unstable-options --profile=fuzz"
        shift
        ;;
      --addrsanitizer)
        export RUSTC_TOOLCHAIN_VERSION="2021-12-22"
        export RUSTFLAGS="-Z sanitizer=address"
        export CARGO_BUILD_TARGET="x86_64-unknown-linux-gnu"
        export PROFILE="debug"
        export CARGO_PROFILE_ARG=""
        shift
        ;;
      --)
        shift
        break
        ;;
      *)
        echo "Invalid function argument"
        exit 1
        ;;
    esac
  done
}

print_configuration() {
  echo -ne "\033[1;37mRunning Tezedge node in ${PROFILE^^} mode"
  if [ -n "$CARGO_BUILD_TARGET" ]; then
      echo -ne "($CARGO_BUILD_TARGET)"
  fi

  if [ -n "$RUSTFLAGS" ]; then
      echo -ne " extra compilation flags : '$RUSTFLAGS'"
  fi
  echo -ne "\e[0m\n"
}

build_all() {
  export SODIUM_USE_PKG_CONFIG=1
  cargo build $CARGO_PROFILE_ARG
}

run_node() {

  # Default light-node commandline arguments:
  #
  # -d | --tezos-data-dir
  TEZOS_DIR=/tmp/tezedge
  # -c | --config-file-path
  CONFIG_FILE=./light_node/etc/tezedge/tezedge.config
  # -B | --bootstrap-db-path
  BOOTSTRAP_DIR=bootstrap_db
  # -i | --identity
  IDENTITY_FILE=./light_node/etc/tezedge/identity.json
  # rest of the commandline arguments will end up here
  args=()

  shift # shift past <MODE>

  # The following loop is used to replace default light-node commandline parameters.
  # It also allows to specify additional commandline parameters.
  # Supports '--arg=val' and '--arg val' syntax of the commandline arguments.
  while [ "$#" -gt 0 ]; do
    case $1 in
      --tezos-data-dir=*)
        TEZOS_DIR="${1#*=}"
        shift
        ;;
      --tezos-data-dir)
        shift
        TEZOS_DIR="$1"
        shift
        ;;
      --config-file=*)
        CONFIG_FILE="${1#*=}"
        shift
        ;;
      --config-file)
        shift
        CONFIG_FILE="$1"
        shift
        ;;
      --bootstrap-db-path=*)
        BOOTSTRAP_DIR="${1#*=}"
        shift
        ;;
      --bootstrap-db-path)
        shift
        BOOTSTRAP_DIR="$1"
        shift
        ;;
      *)
        args+=("$1")
        shift
        ;;
    esac
  done


  # cleanup data directory
  if [ -z "$KEEP_DATA" ]; then
    rm -rf $TEZOS_DIR $BOOTSTRAP_DIR || true
    mkdir -p $TEZOS_DIR $BOOTSTRAP_DIR
  fi

  # protocol_runner needs 'libtezos.so' to run
  export LD_LIBRARY_PATH="${BASH_SOURCE%/*}/tezos/sys/lib_tezos/artifacts:${BASH_SOURCE%/*}/target/$PROFILE"
  # start node
  if [ -z "$CARGO_BUILD_TARGET" ]; then
    PROTOCOL_RUNNER_BINARY=./target/$PROFILE/protocol-runner
  else
    PROTOCOL_RUNNER_BINARY=./target/$CARGO_BUILD_TARGET/$PROFILE/protocol-runner
  fi

  cargo run $CARGO_PROFILE_ARG --bin light-node -- \
            --config-file "$CONFIG_FILE" \
            --tezos-data-dir "$TEZOS_DIR" \
            --identity-file "$IDENTITY_FILE" \
            --bootstrap-db-path "$BOOTSTRAP_DIR" \
            --protocol-runner $PROTOCOL_RUNNER_BINARY "${args[@]}"
}

run_docker() {
  shift # shift past <MODE>

  # build docker
  docker build -t tezedge-run -f ./docker/run/Dockerfile .
  # run docker
  docker run -i -p 9732:9732 -p 18732:18732 -p 4927:4927 -t tezedge-run "$@"
}

run_sandbox() {
  # Default light-node commandline arguments:
  #
  # -d | --tezos-data-dir
  TEZOS_DIR=/tmp/tezedge
  # -B | --bootstrap-db-path
  BOOTSTRAP_DIR=bootstrap_db
  # -n | --network
  NETWORK=sandbox

  # protocol_runner needs 'libtezos.so' to run
  export LD_LIBRARY_PATH="${BASH_SOURCE%/*}/tezos/sys/lib_tezos/artifacts:${BASH_SOURCE%/*}/target/$PROFILE"

  export TEZOS_CLIENT_UNSAFE_DISABLE_DISCLAIMER="Y"

  cargo run $CARGO_PROFILE_ARG --bin sandbox -- \
            --log-level "info" \
            --sandbox-rpc-port "3030" \
            --light-node-path "./target/$PROFILE/light-node" \
            --protocol-runner-path "./target/$PROFILE/protocol-runner" \
            --tezos-client-path "./sandbox/artifacts/tezos-client" "${args[@]}"
}

case $1 in

  node)
    configure_env_variables --debug
    warn_if_not_using_recommended_rust
    print_configuration
    build_all
    run_node "$@"
    ;;

  node-saddr)
    configure_env_variables --debug --addrsanitizer
    warn_if_not_using_recommended_rust
    print_configuration
    build_all
    run_node "$@"
    ;;

  release)
    configure_env_variables --release
    warn_if_not_using_recommended_rust
    print_configuration
    build_all
    run_node "$@"
    ;;

  fuzz)
    rm -f ./target/fuzz/deps/*.gcda
    configure_env_variables --fuzz
    warn_if_not_using_recommended_rust
    print_configuration
    build_all
    run_node "$@"
    ;;

  sandbox)
    configure_env_variables --release
    warn_if_not_using_recommended_rust
    print_configuration
    build_all
    run_sandbox "$@"
    ;;

  docker)
    echo -ne "\033[1;37mRunning Tezedge node in docker\e[0m\n"
    run_docker "$@"
    ;;

  -h|--help)
    help
    ;;

  *)
    echo "run '$0 --help' to get more info"
    ;;

esac

